### ISSUE TITLE
Bug: [no-floating-promises] TypeOrValueSpecifier (allowForKnownSafePromises) doesn't match FastifyRegister

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#10665]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

Following https://github.com/fastify/fastify/pull/5506 -> https://fastify.dev/docs/v5.1.x/Reference/TypeScript#plugins, the [`@typescript-eslint/no-floating-promises` `allowForKnownSafePromises` option](https://typescript-eslint.io/rules/no-floating-promises/#allowforknownsafepromises) should work to allowlist Fastify promises when the [`checkThenables` option](https://typescript-eslint.io/rules/no-floating-promises/#checkthenables) would cause them to report.

However, in https://github.com/typescript-eslint/examples/pull/22, I'm seeing that `FastifyRegister` is still being reported.

```ts
{
  "rules": {
    "@typescript-eslint/no-floating-promises": ["error", {
      "allowForKnownSafePromises": [
        { "from": "package", "name": "FastifyInstance", "package": "fastify" },
        { "from": "package", "name": "FastifyRegister", "package": "fastify" },
        { "from": "package", "name": "SafePromiseLike", "package": "fastify" },
      ]
    }]
  }
}
```

```ts
// imported from "fastify" and "fastify-cookie"
fastify.register(FastifyCookie);
```

```plaintext
  11:1  error  Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
```

I debugged into this a bit and found that:
* In `specifierNameMatches`, the `FastifyRegister` name is correctly being matched ✅ 
* However, in `typeMatchesSpecifier`, the `symbol` is `undefined`

I haven't investigated deeply enough yet to understand why the type wouldn't have a symbol. I don't think this is a user misconfiguration case. The repro is pretty straightforward; its ESlint config uses `projectServices` and its TSConfig has `NodeNext` for both `module` and `moduleResolution`.

Note that Fastify types use `PromiseLike` rather than `Promise`, so the [`checkThenables` option](https://typescript-eslint.io/rules/no-floating-promises/#checkthenables) must be enabled to show the lint complaints.

### Reproduction Repository Link

https://github.com/typescript-eslint/examples/pull/22

### Repro Steps

```shell
git clone https://github.com/typescript-eslint/examples
cd examples
gh pr checkout 22
npm i
cd packages/fastify-floating-promises
npm run lint
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.18.0` |
| `TypeScript`                           | `5.7.2` |
| `ESLint`                               | `9.16.0` |

💖

**Comment by ronami** [comment]
Interesting issue.

The root cause is that the type returned by `register()` isn't `FastifyRegister<...>`, but [an intersection](https://github.com/fastify/fastify/blob/15c91ac05c5560d5d3a56d4cd43d9a224201eda8/types/instance.d.ts#L174) with `FastifyRegister<...>` being one element of it.

Since `register()` doesn't have a uniquely named type, I don't think `TypeOrValueSpecifier` can currently target it.

One potential solution to this is a change on the library's end (which is not optimal, in my opinion).

Alternatively, `TypeOrValueSpecifier` could be adjusted to check each element in an intersection *in addition* to the original symbol of a type. I've tested this locally, and it seems to fix the issue without breaking any existing tests. However, this could sometimes make `TypeOrValueSpecifier` target potentially unwanted types.

Interestingly, [intersections are already somewhat supported](https://github.com/typescript-eslint/typescript-eslint/blob/70f3092d7d4b039317735a0c8abe2229d5678dd6/packages/type-utils/src/typeOrValueSpecifiers/specifierNameMatches.ts#L22-L24), which makes the [`specifierNameMatches` check pass](https://github.com/typescript-eslint/typescript-eslint/blob/70f3092d7d4b039317735a0c8abe2229d5678dd6/packages/type-utils/src/TypeOrValueSpecifier.ts#L176-L178) (for the provided reproduction) but [fails to determine](https://github.com/typescript-eslint/typescript-eslint/blob/70f3092d7d4b039317735a0c8abe2229d5678dd6/packages/type-utils/src/typeOrValueSpecifiers/typeDeclaredInPackageDeclarationFile.ts#L54-L55) if the symbol is in the `fastify` package. Since this is already partly supported, I think the rest of `TypeOrValueSpecifier` should also check each part of an intersection.

+1 from me for `accepting PRs` as a bug.
